### PR TITLE
Testing: use a better way to get a port to the test app

### PR DESCRIPTION
**Pull request by arunoda**
Use http module's `server.listen()` to bind to a random available port.

### CODE REVIEW

**Comment by rauchg** [comment]
Still not very robust. Why not ephemeral ports given by `listen()` ?

**Comment by sedubois** [comment]
Random behavior in tests is nice but can also lead to randomly failing tests sometimes hard to track down (saw that in my previous company). How about testing several startups using a predefined list of ports?

**Comment by rauchg** [comment]
Can you elaborate? How are ephemeral ports error pone?

**Comment by arunoda** [comment]
> Random behavior in tests is nice but can also lead to randomly failing tests sometimes hard to track down

This is not true. We use [portfinder](https://www.npmjs.com/package/portfinder) and it'll give an available port. It starts with 8000.
Here we've randomized that.

**Comment by arunoda** [comment]
> Still not very robust. Why not ephemeral ports given by listen() ?

That's a good idea. I'll update.

**Comment by arunoda** [comment]
@rauchg updated as you suggested.

File: yarn.lock [hunk 1]
```diff
@@ -4290,6 +4296,13 @@ promise@^7.1.1:
...
+promisify-node:
+  version "0.4.0"
+  resolved "https://registry.yarnpkg.com/promisify-node/-/promisify-node-0.4.0.tgz#32803874ec411784e4786c339902a87a179a469c"
+  dependencies:
+    nodegit-promise "~4.0.0"
+    object-assign "^4.0.1"
+
 prr@~0.0.0:
   version "0.0.0"
   resolved "https://registry.yarnpkg.com/prr/-/prr-0.0.0.tgz#1a84b85908325501411853d0081ee3fa86e2926a"
```
Comment (rauchg) [inline yarn.lock]: when you remove a project, you need to re-instantiate the lock. we should be seeing red here, for portmapper

**Review by rauchg — CHANGES_REQUESTED** [review]

File: yarn.lock (hunk 1 above)
Comment (arunoda) [inline yarn.lock]: Yep yep. Need to do it. I use both npm and yarn at the same time. That's a bad habit. 

Will update.

**Comment by sedubois** [comment]
> Can you elaborate?

I was just speaking in a general context. A test which is based on random behavior isn't repeatable. It might start to fail sometimes but not every time, might be quite unsettling. A test (explicitly or not) has 3 parts: the "given", the "when" and the "then". It would be quite helpful to know the "given" just by reading the source code.

**Comment by rauchg** [comment]
But the randomness of the port has no impact in the outcome, correct? Specially considering that ephemeral ports (via `listen()`) always give us a free port.

File: yarn.lock (hunk 1 above)
Comment (arunoda) [inline yarn.lock]: @rauchg done.

**Comment by sedubois** [comment]
If the randomness has no impact on the outcome, why do you use it? Why make this PR in that case?

**Comment by rauchg** [comment]
Because let's say you have something running on port 8000, the tests would have failed for the _wrong_ reason

**Comment by sedubois** [comment]
I think I just haven't understood why that port would be randomly unavailable in the first place. Normally the point of a test env is to offer a reproducible context.

Anyways can just discard my comments, it's just that the subject of randomness triggered some sensitive memories from a past life at F-Secure 😄

**Comment by sedubois** [comment]
I hope the port isn't randomly unavailable in the test automation env? So it's on some developers' laptops that the issue arises? Then other option than randomness might be to allow the dev to configure the port when running tests (with a sensible default).

**Comment by arunoda** [comment]
I think here we started to randomize the base port. But now we are doing something else.
We let node's http module to pick an available port and use for the test app.

**Comment by sedubois** [comment]
Other issue with random port is that it might hide the fact that another instance of the tests might have been stuck and left running elsewhere on the machine.

**Comment by arunoda** [comment]
@sedubois we always run test in parallel and Jest know how to report them well.

> it might hide the fact that another instance of the tests might have been stuck and left running elsewhere on the machine.

Let's ignore this case for now. Let's do things one by one.
We can create a new issue and continue.

This is a good improvement compared with the current setup as we longer using  `portfinder`.

**Comment by sedubois** [comment]
👍

**Comment by rauchg** [comment]
> I just hadn't understood why that port would be randomly unavailable in the first place. Normally the point of a test env is to offer a reproducible context.

@sedubois sure but we also run tests upon `prepush`, `prepublish`, so the tests have to be useful on developer machines

**Comment by rauchg** [comment]
Thanks for your input @sedubois, always much appreciated

**Comment by timneutkens** [comment]
@arunoda updated the branch https://github.com/arunoda/next.js/pull/1.

**Review by timneutkens — APPROVED** [review]
Looks good 👍 @arunoda ready for merge?

**Comment by arunoda** [comment]
I think this is good to merge.

**Comment by timneutkens** [comment]
Awesome ❤️

File: package.json [hunk 2]
```diff
@@ -104,9 +104,6 @@
     "jest-cli": "^18.0.0",
     "node-fetch": "^1.6.3",
     "nyc": "^10.0.0",
-    "portfinder": "^1.0.10",
-    "react": "15.4.2",
-    "react-dom": "15.4.2",
```
Comment (nkzawa) [inline package.json]: Please don't remove `react` and `react-dom`. Tests fail when you did clean install.

Comment (timneutkens) [inline package.json]: Ah right, my fault when resolving merge conflicts.

**Pull request merged**